config-linux.md

```diff
@@ -115,93 +107,59 @@ There is a limit of 5 mappings which is the Linux kernel hard limit.
 
 ## Devices
 
-`devices` is an array specifying the list of devices to be created in the container.
+`devices` is an array specifying the list of devices that MUST be available in the container.
+The runtime may supply them however it likes (with [mknod][mknod.2], by bind mounting from the runtime mount namespace, etc.).
 
 The following parameters can be specified:
 
-* **`type`** *(char, required)* - type of device: `c`, `b`, `u` or `p`. More info in `man mknod`.
-
-* **`path`** *(string, optional)* - full path to device inside container
-
-* **`major, minor`** *(int64, required)* - major, minor numbers for device. More info in `man mknod`. There is a special value: `-1`, which means `*` for `device` cgroup setup.
-
-* **`permissions`** *(string, optional)* - cgroup permissions for device. A composition of `r` (*read*), `w` (*write*), and `m` (*mknod*).
-
-* **`fileMode`** *(uint32, optional)* - file mode for device file
-
-* **`uid`** *(uint32, optional)* - uid of device owner
-
-* **`gid`** *(uint32, optional)* - gid of device owner
-
-**`fileMode`**, **`uid`** and **`gid`** are required if **`path`** is given and are otherwise not allowed.
+* **`type`** *(char, required)* - type of device: `c`, `b`, `u` or `p`.
+  More info in [mknod(1)][mknod.1].
+* **`path`** *(string, required)* - full path to device inside container.
+* **`major, minor`** *(int64, required)* - [major, minor numbers][devices] for the device.
```

Comment:
On Wed, Jan 27, 2016 at 01:13:52PM -0800, Vish Kannan wrote:
> > +* **`major, minor`** *(int64, required)* - [major, minor numbers][devices] for the device.
>
> This field should not be specified if the type is `p`. Hence it must
> be optional

It's hard to create a block or character special file if either of
those is left off.  How about:

* **`major, minor`** *(int64, required unless **`type`** is `p`)* - …

Obviously, that means they need to be optional in the Go type
definition, and we'd need more complicated validation logic to check
that condition.  But I think:

  {
    "path": "/dev/fuse",
    "type": "c",
    "fileMode": 0666,
    "uid": 0,
    "gid": 0
  }

is pretty clearly an invalid entry.